lib/cylc/network/scan.py

```diff
@@ -0,0 +1,257 @@
+#!/usr/bin/env python3
+
+# THIS FILE IS PART OF THE CYLC SUITE ENGINE.
+# Copyright (C) 2008-2019 NIWA & British Crown (Met Office) & Contributors.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Port scan utilities."""
+
+import asyncio
+import os
+from pwd import getpwall
+import re
+import sys
+import socket
+
+from cylc.cfgspec.glbl_cfg import glbl_cfg
+import cylc.flags
+from cylc.hostuserutil import is_remote_host, get_host_ip_by_name
+from cylc.network.client import (
+    SuiteRuntimeClient, ClientError, ClientTimeout)
+from cylc.suite_srv_files_mgr import (
+    SuiteSrvFilesManager, SuiteServiceFileError)
+
+DEBUG_DELIM = '\n' + ' ' * 4
+INACTIVITY_TIMEOUT = 10.0
+MSG_QUIT = "QUIT"
+MSG_TIMEOUT = "TIMEOUT"
+SLEEP_INTERVAL = 0.01
+
+
+def async_map(coroutine, iterator):
+    """Map iterator iterator onto a coroutine.
+
+    * Yields results in order as an when they are ready
+    * Slow workers can block.
+
+    Args:
+        coroutine (asyncio.coroutine):
+            I.E. an async function.
+        iterator (iter):
+            Should yield tuples to be passed into the coroutine.
+
+    Yields:
+        list - List of results
+
+    Example:
+        >>> async def square(number): return number ** 2
+        >>> generator = async_map(square, ((i,) for i in range(5)))
+        >>> list(generator)
+        [0, 1, 4, 9, 16]
+
+    """
+    loop = asyncio.get_event_loop()
+
+    awaiting = []
+    for ind, args in enumerate(iterator):
+        task = loop.create_task(coroutine(*args))
+        task.ind = ind
+        awaiting.append(task)
+
+    index = 0
+    buff = []
```

Comment:
Would it be better to use a `set` or `dict` here? Can we rename `buff` to `tasks` or `completed_tasks` here?

Reply:
* Will rename.
* `buff` needs to maintain order so `set` wouldn't work.
* items don't have associated keys so `dict` wouldn't be the first choice.

I must be missing something here, what's your logic?

Reply:
Since the logic appears to be doing a lot of `.extend` and `.remove`. More natural to use `dict`?